### PR TITLE
Unit tests, Travis CI and Coveralls integration

### PR DESCRIPTION
**Pull request by cevitcejbo**
Added tests for all public methods of RKNotificationHub using Specta/Expecta, FBSnapshots and Expecta+Snapshots during [CocoaPods Test Jam](https://cocoapods.org/testjam).
Currently Travis CI and Coveralls integration uses my GitHub account. Feel free to turn them on yourself as I don't have admin rights on the original repo. See https://travis-ci.org and https://coveralls.io.

Current coverage is 88%. Note that `- (CGRect)nextRectWithDiameter:(float)diameter` method is not covered since it's not used anywhere (please review this). Demo ViewController class is also not covered and it's included in stats since it's located in the same folder as RKNotificationHub.m.

### CODE REVIEW

**Comment by cwRichardKim** [comment]
hey i'm really sorry about not responding, I turned off notifications for this by accident. checking out the code now

**Comment by cwRichardKim** [comment]
removed nextRectWithDiameter, nice catch

**Comment by cwRichardKim** [comment]
I'm a little confused as to what this PR achieves.  Would you mind explaining a little?

**Comment by cevitcejbo** [comment]
The PR doesn't change any existing behaviour of the library. But it does cover almost all the code with unit tests, thus facilitating possible further changes, refactoring efforts and in a certain way documenting current behaviour using specs. I was using Specta/Expecta for test cases and FBSnapshotTestCase for snapshot testing (here's a good source on testing techniques in iOS: http://www.objc.io/issue-15/). I've added tests to the test target of the demo project (RKNotificationHub.xcodeproj). Please note that you need to perform "pod install" and then open generated RKNotificationHub.xcworkspace to be able to run the tests (⌘+u in Xcode).
Moreover, the PR enables automated builds of the test target on Travis CI on every change pushed to the repository. In addition, generated test coverage stats are sent from Travis CI to the Coveralls service making it possible to track changes in code coverage stats.
Please let me know if you would need any further details on the PR.

**Comment by cwRichardKim** [comment]
Ok so I'm new to this kind of thing, but from my experience, testing is only really needed when you're working on an expansive project that has many moving parts.  I've heard of specta, but why would one need a testing framework for a pretty basic UI-based repository? I'm unfamiliar with Travis, but again, I'm struggling to see why it would be necessary for something like this.

What benefit does this have for the end user? (the developer using the pod).  It would make it a little less readable in my opinion.

File: RKNotificationHubTests/RKNotificationHubTests.m [hunk 1]
```diff
@@ -1,40 +1,277 @@
...
+        it(@"should increase count", ^{
+            [subject increment];
+            expect([subject count]).to.equal(21);
+        });
+    });
+    
+    describe(@"incrementBy:", ^{
+        it(@"should increase count by given value", ^{
+            [subject incrementBy:10];
+            expect([subject count]).to.equal(30);
```
Comment (cwRichardKim) [inline RKNotificationHubTests/RKNotificationHubTests.m]: shouldn't this be 31?

Comment (cwRichardKim) [inline RKNotificationHubTests/RKNotificationHubTests.m]: shouldn't this be 31?

File: RKNotificationHubTests/RKNotificationHubTests.m [hunk 2]
```diff
@@ -1,40 +1,277 @@
...
+        });
+    });
+    
+    describe(@"decrementBy:", ^{
+        it(@"should decrease count by given value", ^{
+            [subject decrementBy:15];
+            expect([subject count]).to.equal(5);
+        });
+        
+        it(@"should make count negative if given value is larger than the count", ^{
```
Comment (cwRichardKim) [inline RKNotificationHubTests/RKNotificationHubTests.m]: wow you really turned this repo inside out didn't you

File: RKNotificationHubTests/RKNotificationHubTests.m (hunk 1 above)
Comment (cevitcejbo) [inline RKNotificationHubTests/RKNotificationHubTests.m]: Good question. This is actually a feature of BDD-style syntax: sibling blocks are independent of each other. The order of calls for the first two tests from "changing the count" is like this:
1. before block -> describe "increment" block -> it "should increase count" block
2. before block -> describe "incrementBy:" block -> it "should increase count by given value" block
A new RKNotificationHub object is created with count 20 for each run in the "before" block.

File: RKNotificationHubTests/RKNotificationHubTests.m (hunk 2 above)
Comment (cevitcejbo) [inline RKNotificationHubTests/RKNotificationHubTests.m]: Yes, I've spotted this behaviour when I was playing with the library and decided to make it explicit by adding it to the spec. This example illustrates how specs can be useful for the end user (the developer, as you've stated earlier). Besides being a unit testing facility, the spec works like a documentation and a set of usage samples of the library and one can simply look through the spec to better understand its behaviour and spot not-so-obvious cases. As for this particular case, it shows that `decrement` and `decrementBy:` behave slightly different when current count is 0 (although I'm not sure that you have actually intended to have this distinction from the very beginning :) ).

**Comment by cevitcejbo** [comment]
>Ok so I'm unfamiliar with this kind of thing, but from my experience, testing is only really needed when you're working on an expansive project that has many moving parts. I've heard of specta, but why would one need a testing framework for a pretty basic UI-based repository?

I would agree that unit tests are probably much more valuable when you are dealing with a big library with thousands and thousands lines of a code especially when the code was contributed by different people. But actually any library that have any kind of behaviour can benefit from being covered with tests. Even those libraries that are more UI-oriented; their views can be tested using [snapshot testing techniques](https://github.com/facebook/ios-snapshot-test-case).

>I'm unfamiliar with Travis, but again, I'm struggling to see why it would be necessary for something like this (also I don't think I understand what it does. The website isn't very clear).

It's a great automated tool that performs builds and runs all tests for the repo thus checking the status of the build on every change that was pushed to the repo. If someone sends a PR, you will be able to see if those changes break the build or make any test fail (or even if it changes the test coverage stats in case you also have Coveralls integrated).

>What benefit does this have for the end user? (the developer using the pod). Or is it mostly just for me for when I further development of the pod?

Unit testing facilitates further changes (improvements or refactoring of the existing features, new features) by providing a simple way to check if everything still behaves as expected. This is especially helpful for other contributors that might not know all the implicit intricacies of the library implementation and preventing them from changing behaviours that they haven't intended to change.
Also, BDD-styled specs serve as a kind of a documentation and library usage samples. It also provides a clear context of a failing scenario if any test fails.
One more benefit for the end user (developer) is that test coverage gives more confidence about the reliability of the code and thus making it safer to use it in production. Code coverage stats from the Coveralls banner in the Readme file even gives a numeric proof of that :)

**Comment by cevitcejbo** [comment]
This CocoaPods blog post includes a list of pull requests that were made during the Test Jam for other popular pods besides RKNotificationHub: http://blog.cocoapods.org/Jamming-2015/. Note that many of them are also UI-based.

**Comment by cwRichardKim** [comment]
Thanks for the clear response, one last question: for the snapshot technique, this would have to be changed if I ever change the default aesthetic, correct?

**Comment by cevitcejbo** [comment]
Yes, in this case you need to update the snapshots to test against. This is really simple, just replace the calls to `haveValidSnapshot() / haveValidSnapshotNamed(...)` with `recordSnapshot() / recordSnapshotNamed(...)` in corresponding test cases, then run the tests to update snapshot files. Return back to `haveValidSnapshot() / haveValidSnapshotNamed (...)` once you're done. There's also a handy Xcode plugin available for quick browsing through all failed snapshot tests and override old snapshots with updated versions with a single tap: https://github.com/orta/snapshots.

**Comment by cwRichardKim** [comment]
sounds good.  Last last question (sorry haha). So you said I have to run pod install, does this mean everyone is required to have more dependencies when using this repo or is it just for the test / demo project.  Also looks like I should set up travis and coverall accounts huh

**Comment by cevitcejbo** [comment]
Installing RKNotificationHub via using cocoa pods doesn't require more dependencies. But this was a good point anyway: simply checking out the repo and running the demo project required it. I've updated the PR and disabled test target when building for running, so this is no longer the case. The user would need to pod install for the demo project only for running tests while the main target doesn't have any extra dependencies.
As for Travis and Coveralls, the setup is pretty basic. Go to https://travis-ci.org and https://coveralls.io, login using your GitHub account (it's free for open source projects) on both and enable RKNotificationHub repo for both. After any change pushed to the repo (for example, merging PR would do), GitHub would automatically trigger a build script on Travis, which in turn will send code coverage info to Coveralls. The last thing would be to update Travis and Coveralls banner links in the Readme file once you are ready with the integration. Let me know if you have any trouble with those services :)

**Comment by cwRichardKim** [comment]
thanks, this has been really informative. Merging!

**Comment by cwRichardKim** [comment]
I'm getting this error:
diff: /../Podfile.lock: No such file or directory
diff: /Manifest.lock: No such file or directory
error: The sandbox is not in sync with the Podfile.lock. Run 'pod install' or update your CocoaPods installation.

**Comment by cwRichardKim** [comment]
I reverted this for now until the error is figured out.  That's my bad for not checking out the branch first.  I think it should definitely be a requirement that someone can without cocoapods setup should be able to open and run the project.

**Comment by cevitcejbo** [comment]
Sorry you had to revert this. Actually those errors are related to the test target only and I'm able to run the main target without any extra work (please try this on your side and confirm). But, I would agree that it's not clear for the developer that those errors can be ignored as long as they don't use test target.
I would suggest 2 ways of resolving this:
1. Remove Pods and *.xcworkspace from .gitignore and push pods sources to the repo, so there's no need to have an environment with cocoa pods installed; all dependencies would be packaged right within the repo (here's a helpful note on this approach vs. keeping it in .gitignore: https://guides.cocoapods.org/using/using-cocoapods.html#should-i-check-the-pods-directory-into-source-control).
2. Create a separate Xcode project inside the repo solely for testing. This project would require "pod install", but the original demo project would remain as it was.
Let's decide what option to use and I will amend my PR.

**Pull request merged**